examples/win-offsets.c

```diff
@@ -51,76 +79,241 @@ void sigint_handler()
+    if (VMI_FAILURE == vmi_read_addr(vmi, &ctx, &kthread)) {
+        dp("Failed to get current KTHREAD from GS_BASE\n");
+        goto failed;
+    }
+
+    /*
+     * Find _EPROCESS of currently running thread.
+     */
+    addr_t pkprocess = kthread + offset_kthread_process;
+
+    ctx.addr = pkprocess;
+    if (VMI_FAILURE == vmi_read_addr(vmi, &ctx, &eprocess)) {
+        dp("Failed to get EPROCESS from KTHREAD\n");
+        goto failed;
+    }
+
+    /*
+     * Find PID of currently running process.
+     */
+    addr_t pid_ptr = eprocess + offset_eprocess_uniqueprocessid;
+
+    ctx.addr = pid_ptr;
+    if (VMI_FAILURE == vmi_read_addr(vmi, &ctx, &pid)) {
+        dp("Failed to get PID from EPROCESS\n");
+        goto failed;
+    }
+
+    /*
+     * Check if we've caught System process. This would ensure that
+     * current DTB for VCPU=0 contains all necessary kernel mappings.
+     * These mappings may not be present or be incomplete in other processes
+     * due to KPTI.
+     */
+    if (pid != 4) {
+        dp("Current PID=%llx, skip until we reach PID=4\n", (unsigned long long)pid);
+        goto failed;
+    }
+
+    dp("Stopped inside system process, VCPU %lu\n", event->vcpu_id);
+    find_pid4_success = 1;
+    found_vcpu_id = event->vcpu_id;
+
+    /*
+     * Remove CR3 event and leave the VM paused inside System process,
+     * to make it easy for LibVMI to detect all necessary offsets.
+     */
+    dp("Cleared event\n");
+    vmi_pause_vm(vmi);
+    vmi_clear_event(vmi, event, NULL);
+    return VMI_EVENT_RESPONSE_NONE;
+
+failed:
+    return VMI_EVENT_RESPONSE_NONE;
+}
+
+void show_usage(char *arg0)
+{
+    fprintf(stderr, "Usage: %s name|domid <domain name|domain id> -r <rekall profile> [-v] [-k]\n", arg0);
+    fprintf(stderr, "    [-v]   optional, enable verbose mode\n");
+    fprintf(stderr, "    [-k]   optional, only print KPGD value\n");
+}
+
 int main(int argc, char **argv)
 {
     vmi_mode_t mode;
     int rc = 1;
 
-    /* this is the VM that we are looking at */
-    if (argc != 5) {
-        printf("Usage: %s name|domid <domain name|domain id> -r <rekall profile>\n", argv[0]);
-        return 1;
-    }   // if
-
     void *domain;
     uint64_t domid = VMI_INVALID_DOMID;
     uint64_t init_flags = 0;
+    int only_output_kpgd = 0;
+
+    char *rekall_profile = NULL;
+    char c;
+
+    while ((c = getopt (argc, argv, "kvr:")) != -1)
+        switch (c) {
+            case 'k':
+                only_output_kpgd = 1;
+                break;
+            case 'v':
+                enable_debug = 1;
+                break;
+            case 'r':
+                rekall_profile = optarg;
+                break;
+            default:
+                show_usage(argv[0]);
+                return 1;
+        }
+
+    if (argc - optind != 2) {
+        show_usage(argv[0]);
+        return 1;
+    }
 
-    if (strcmp(argv[1],"name")==0) {
-        domain = (void*)argv[2];
+    if (strcmp(argv[optind],"name")==0) {
```

Comment:
Can we switch this to just use standard `getopt` flags? Like `--domain` and `--domid`?

Reply:
done